AlphaWallet/Core/CoinTicker/CoinTickersResolver.swift

```diff
@@ -0,0 +1,243 @@
+
+    func coinTickers(enabledObject: [TokenMappedToTicker]) -> Promise<Void>
+    func chartHistory(force: Bool, period: ChartHistoryPeriod, for key: AddressToPRCServerKey) -> Promise<ChartHistory>
+}
+
+class CoinTickersResolver: CoinTickersResolverType {
+
+    private enum AnyError: Error {
+        case alreadyFetchingPrices
+        case minTimeHasntPassYet
+    }
+
+    private static var coinTickerIds: [CoinTickerId] = []
+
+    private var isFetchingPrices = false
+    private static let queue: DispatchQueue = DispatchQueue(label: "com.DigitalBitsCoinTickersResolver.updateQueue")
+
+    private let provider: MoyaProvider<AlphaWalletService>
+    private let server: RPCServer
+    private let config: Config
+    private (set) var tickers: [AddressToPRCServerKey: CoinTicker] = [:]
+    private (set) var history: [CoinTicker: [ChartHistoryPeriod: ChartHistory]] = [:]
+
+    init(provider: MoyaProvider<AlphaWalletService>, server: RPCServer, config: Config) {
+        self.provider = provider
+        self.server = server
+        self.config = config
+    }
+
+    private func fetchCoinTikerIds() -> Promise<[CoinTickerId]> {
+        if CoinTickersResolver.coinTickerIds.isEmpty {
+            return firstly {
+                provider.request(.coins(config: config))
+            }.map(on: CoinTickersResolver.queue, { response -> [CoinTickerId] in
+                return try response.map([CoinTickerId].self, using: JSONDecoder())
+            }).map(on: CoinTickersResolver.queue, { ids in
+                CoinTickersResolver.coinTickerIds = ids
+
+                return ids
+            }).recover { _ -> Promise<[CoinTickerId]> in
+                return .value([])
+            }
+        } else {
+            return .value(CoinTickersResolver.coinTickerIds)
+        }
+    }
+
+    private var fetchTickersIsAvailable: Bool {
+        switch server {
+        case .main, .xDai:
+            return true
+        case .kovan, .ropsten, .rinkeby, .poa, .sokol, .classic, .callisto, .goerli, .artis_sigma1, .artis_tau1, .binance_smart_chain, .binance_smart_chain_testnet, .custom, .heco, .heco_testnet, .fantom, .fantom_testnet, .avalanche, .avalanche_testnet, .polygon, .mumbai_testnet:
+            return false
+        }
+    }
+
+    private struct MappedCoinTickerId: Hashable {
+        let coinTickerId: String
+        let contractAddress: AlphaWallet.Address
+        let server: RPCServer
+
+        init(id: String, contractAddress: AlphaWallet.Address, server: RPCServer) {
+            self.coinTickerId = id
+            self.contractAddress = contractAddress
+            self.server = server
+        }
+    }
+
+    func coinTickers(enabledObject: [TokenMappedToTicker]) -> Promise<Void> {
+        Promise<Void> { seal in
+            coinTickersPromise(enabledObject: enabledObject).done { [weak self] tickers in
+                self?.tickers = tickers
+                self?.loadChartHistoryForResolverTickers()
```

Comment:
Does this mean that this shouldn't be public, since it can be called externally (again)?

```
    func chartHistory(force: Bool, period: ChartHistoryPeriod, for key: AddressToPRCServerKey) -> Promise<ChartHistory> {
```